man/fpsync.1
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.\" Copyright (c) 2015-2026 Ganael LAPLANCHE <ganael.laplanche@martymac.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd January 27, 2015
.Dt FPSYNC 1
.Os
.Sh NAME
.Nm fpsync
.Nd Synchronize directories in parallel using fpart and an external tool
.Sh SYNOPSIS
.Nm
.Op Fl p
.Op Fl n Ar jobs
.Op Fl w Ar wrks
.Op Fl m Ar tool
.Op Fl T Ar path
.Op Fl f Ar files
.Op Fl s Ar size
.Op Fl E
.Op Fl o Ar toolopts
.Op Fl O Ar fpartopts
.Op Fl S
.Op Fl t Ar tmpdir
.Op Fl d Ar shdir
.Op Fl M Ar mailaddr
.Op Fl Z
.Op Fl v
.Pa src_dir/
.Pa dst_url/
.Nm
.Fl l
.Nm
.Fl r Ar runid
.Op Fl R
.Cm [OPTIONS...]
.Nm
.Fl a Ar runid
.Nm
.Fl D Ar runid
.Sh DESCRIPTION
The
.Nm
tool synchronizes directories in parallel using
.Xr fpart 1
and
.Xr rsync 1 ,
.Xr cpio 1 ,
.Xr pax 1
or
.Xr tar 1 .
It computes subsets of
.Pa src_dir/
and spawns jobs to synchronize them to
.Pa dst_url/ .
.sp
Synchronization jobs can be executed either locally or remotely (using SSH
workers, see option
.Fl w )
and are executed on-the-fly while filesystem crawling goes on.
This makes
.Nm
a good tool for migrating large filesystems.
.Sh COMMON OPTIONS
.Bl -tag -width indent
.It Fl t Ar tmpdir
Set
.Nm
temporary directory to
.Ar tmpdir .
This directory remains local and does not need to be shared amongst SSH workers
when using the
.Fl w
option.
Default:
.Pa /tmp/fpsync
.It Fl d Ar shdir
Set
.Nm
shared directory to
.Ar shdir .
This option is mandatory when using SSH workers and set by default to
.Ar tmpdir
when running locally.
The specified directory must be an absolute path ; it will be used to handle
communications with SSH hosts (sharing partitions and log files) and, as a
consequence, must be made available to all participating hosts (e.g. through a
r/w NFS mount), including the master one running
.Nm .
.It Fl M Ar mailaddr
Send an e-mail to
.Ar mailaddr
after a run.
Multiple -space-separated- addresses can be specified.
That option requires the
.Xr mail 1
client to be installed and configured on the master host running
.Nm .
.It Fl Z
Prefix logs with human-readable dates instead of timestamps.
.It Fl v
Verbose mode.
Can be be specified several times to increase verbosity level.
.It Fl h
Print help
.It Fl V
Print version
.El
.Sh SYNCHRONIZATION OPTIONS
.Bl -tag -width indent
.It Fl m Ar tool
External copy
.Ar tool
used to synchronize files.
Currently supported tools are:
.Sy rsync ,
.Sy cpio ,
.Sy pax ,
.Sy tar ,
and
.Sy tarify .
Default:
.Sy rsync .
Special tool
.Sy tarify
generates tarballs into destination directory.
.It Fl T Ar path
Specify absolute
.Ar path
of copy tool (guessed by default).
If you force a specific
.Ar path ,
the copy tool must be present at that
.Ar path
on each worker.
That
.Ar path
cannot be changed when resuming a run.
.It Fl f Ar files
Transfer at most
.Ar files
files or directories per sync job.
.Sy 0
means unlimited but you must at least specify one file or size limit.
Default:
.Sy 2000
.It Fl s Ar size
Transfer at most
.Ar size
bytes per sync job.
.Sy 0
means unlimited but you must at least specify one file or size limit.
You can use a human-friendly unit suffix here (k, m, g, t, p).
.br
Default:
.Sy 4294967296
(4 GB)
.It Fl E
Work on a per-directory basis (rsync tool only).
In that mode,
.Nm
works with lists of directories instead of files.
That mode may generate coarse-grained lists but enables
.Xr rsync 1 's
.Cm --delete
option by default (
.Sy WARNING ! ! !
), making it a good candidate for a final (cleaning) pass after several
incremental passes using standard (file) mode.
When option
.Fl E
is specified twice, it enables 'aggressive' mode which isolates erroneous
directories and enables recursive synchronization for them.
This advanced mode can be useful to try to overcome transcient errors such as
Linux SMB client deferring
.Fn opendir
calls to support compound SMB requests.
.It Fl o Ar toolopts
Override default
.Xr rsync 1 ,
.Xr cpio 1 ,
.Xr pax 1
or
.Xr tar 1
options with
.Ar toolopts .
Use this option with care as certain options are incompatible with a parallel
usage (e.g. rsync's
.Cm --delete ) .
Default for rsync:
.Dq -lptgoD -v --numeric-ids .
Empty for cpio, pax, tar and tarify.
.It Fl O Ar fpartopts
Override default
.Xr fpart 1
options with
.Ar fpartopts .
Options and values must be separated by a pipe character.
.br
Default:
.Dq -x|.zfs|-x|.snapshot*|-x|.ckpt .
.It Fl S
Sudo mode.
Use
.Xr sudo 8
for filesystem crawling and synchronizations.
.It Pa src_dir/
Source directory.
It must be absolute and available on all participating hosts (including the
master one, running
.Nm ) .
.It Pa dst_url/
Destination directory or URL (rsync tool only).
If a remote URL is provided, it must be supported by
.Xr rsync 1 .
All participating workers must be able to reach that target.
.El
.Sh JOB HANDLING AND DISPATCHING OPTIONS
.Bl -tag -width indent
.It Fl n Ar jobs
Start
.Ar jobs
concurrent sync jobs (either locally or remotely, see below) per run.
Default:
.Sy 2
.It Fl w Ar wrks
Use remote SSH
.Ar wrks
to synchronize files.
Synchronization jobs are executed locally when this option is not set.
.Ar wrks
is a space-separated list of login@machine connection strings and can be
specified several times.
You must be allowed to connect to those machines using a SSH key to avoid user
interaction.
.El
.Sh RUN HANDLING OPTIONS
.Bl -tag -width indent
.It Fl p
Prepare mode: prepare, test synchronization environment, start
.Xr fpart 1
and create partitions but do
.Sy not
actually start transfers.
That mode can be used to create a run that can then be resumed using option
.Fl r .
.It Fl l
List previous runs and their status.
.It Fl r Ar runid
Resume run
.Ar runid
and restart synchronizing remaining partitions from a previous run.
.Ar runid
is displayed when using verbose mode (see option
.Fl v )
or prepare mode (option
.Fl p )
and can be retrieved afterwards using option
.Fl l .
Note that filesystem crawling is skipped when resuming a previous run.
As a consequence, options
.Fl m ,
.Fl f ,
.Fl s ,
.Fl E ,
.Fl o ,
.Fl O ,
.Fl S ,
.Pa src_dir/ ,
and
.Pa dst_url/
are ignored.
.It Fl R
Replay mode: when using option
.Fl r ,
force re-synchronizing run's all partitions instead of remaining ones only.
That mode can be useful to skip filesystem crawling when you have to replay a
final pass several times and you know directory structure has not changed in
the meantime (you may miss files if you use replay mode with a standard,
file-based, run).
.It Fl a Ar runid
Archive run
.Ar runid
(including partition files, logs, queue and work directories) to
.Ar tmpdir .
That option requires the
.Xr tar 1
client to be installed on the master host running
.Nm .
.It Fl D Ar runid
Delete run
.Ar runid
(including partition files, logs, queue and work directories).
.El
.Sh RUNNING FPSYNC
Each
.Nm
run generates a unique
.Ar runid ,
which is displayed in verbose mode (see option
.Fl v )
and within log files.
You can use that
.Ar runid
to resume a previous run (see option
.Fl r ) .
.Nm
will then restart synchronizing data from the parts that were being synchronized
at the time it stopped.
.sp
This unique feature gives the administrator the ability to stop
.Nm
and restart it later, without having to restart the whole filesystem crawling
and synchronization process.
Note that resuming is only possible when filesystem crawling step has finished.
.sp
During synchronization, you can press CTRL-C to interrupt the process.
The first CTRL-C prevents new synchronizations from being submitted and the
process will wait for current synchronizations to be finished before exiting.
If you press CTRL-C again, current synchronizations will be killed and
.Nm
will exit immediately.
When using option
.Fl E
to enable directory mode and rsync's
.Cm --delete
option, keep in mind that killing rsync processes may lead to a situation where
certain files have been updated and others not deleted yet (because the deletion
process is postponed using rsync's
.Cm --delete-after
option).
.sp
On certain systems, CTRL-T can be pressed to get the status of current and
remaining parts to be synchronized.
This can also be achieved by sending a SIGINFO (signal 29) to the
.Nm
process.
.sp
Whether you use verbose mode or not, everything is logged within
.Pa shdir/log/ .
.Sh EXAMPLES
Here are some examples:
.Bl -tag -width indent
.It Li "fpsync -n 4 /usr/src/ /var/src/"
.sp
Synchronizes
.Pa /usr/src/
to
.Pa /var/src/
using 4 local jobs.
.It Li "fpsync -n 4 -f 10000 -s 10g -m cpio /usr/src/ /var/src/"
.sp
The same, using
.Xr cpio 1
instead of
.Xr rsync 1
and partitions containing up to 10000 files and 10 GiB of data.
.It Li "fpsync -n 2 -w login@machine1 -w login@machine2 -d /mnt/fpsync /mnt/src/ /mnt/dst/"
.sp
Synchronizes
.Pa /mnt/src/
to
.Pa /mnt/dst/
using 2 concurrent jobs executed remotely
on 2 SSH workers (machine1 and machine2).
The shared directory is set to
.Pa /mnt/fpsync
and mounted on the machine running
.Nm ,
as well as on machine1 and machine2.
The source directory
.Pa ( /mnt/src/ )
is also available on those 3 machines, while the destination directory
.Pa ( /mnt/dst/ )
is mounted on SSH workers only (machine1 and machine2).
.It Li "fpsync -n 1 -s 20g -vv -E -O '-x|./tmp/' /data/ /mnt/backup/"
.sp
Synchronizes
.Pa /data/
to
.Pa /mnt/backup/
using a single local job (to benefit from fpsync's fast synchronization
startup, as well as its status and resuming facilities).
Option
.Fl E
is enabled, which means fpsync works on a per-directory basis and enables
.Xr rsync 1 's
.Cm --delete
option.
Each job (partition) has a maximum size of 20 GiB.
Finally, contents from '/data/tmp/' is excluded (note that the directory entry
itself is created anyway, as it belongs to its parent folder, but remains
empty).
.El
.Sh LIMITATIONS
Parallelizing
.Xr rsync 1
can make several options not usable, such as
.Cm --delete .
If your source directory is live while
.Nm
is running, you will have to delete extra files from destination directory.
This is traditionally done by using a final -offline-
.Xr rsync 1
pass that will use this option, but you can also use
.Nm
and option
.Cm -E
to perform the same task using several workers.
.sp
.Nm
enqueues synchronization jobs on disk, within the
.Pa tmpdir/queue
directory.
Be careful to host this queue on a filesystem that can handle fine-grained
mtime timestamps (i.e. with a sub-second precision) if you want
the queue to be processed in order when
.Xr fpart 1
generates several jobs per second.
On
.Fx ,
.Xr VFS 9
timestamps' precision can be
tuned using the 'vfs.timestamp_precision' sysctl.
See
.Xr vfs_timestamp 9 .
.sp
Contrary to
.Xr rsync 1 ,
.Nm
enforces the final '/' on the source directory.
It means that directory
.Sy contents
are synchronized, not the source directory itself (i.e. you will not get a
subdirectory of the name of the source directory in the target directory after
synchronization).
.sp
Before starting filesystem crawling,
.Nm
changes its current working directory to
.Pa src_dir/
and generates partitions containing
.Sy relative
paths (all starting with './').
This is important to keep in mind when modifying
.Ar toolopts
or
.Ar fpartopts
dealing with file or directory paths.
.sp
If you want to run
.Nm
in the background, your shell must support enabling job control without a tty
in non-interactive mode.
As of this writing, Dash does not support that feature and will send a SIGTTIN
to the whole process group, suspending fpsync execution.
.Sh SEE ALSO
.Xr cpio 1 ,
.Xr fpart 1 ,
.Xr mail 1 ,
.Xr pax 1 ,
.Xr rsync 1 ,
.Xr tar 1 ,
.Xr sudo 8
.Sh AUTHOR, AVAILABILITY
Fpsync has been written by
.An Gana\(:el LAPLANCHE
and is available under the BSD
license on
.Lk https://contribs.martymac.org
.Sh BUGS
No bug known (yet).


Changelog
2021/??/??, 1.3.1 ('???') :
    - fpsync: allow special characters and white spaces within fpart options
      Allows something like: fpsync ... -O "-y|foo bar|-y|*.txt" /src/ /dst/
      (fixes GH issue #33)
2021/05/08, 1.3.0 ('Cleaning room') :
    - fpart: align get_size() and init_file_entries() behaviour and take
      exclusions into account when computing a directory size.
    - fpart: fix directory size computation when using options -y/-Y.
      Before that change, -D and -E would have returned a size of 0 for a
      directory included through options -y/-Y.
    - fpart: fix detection of an unknown option
    - fpart: various fixes on Darwin
    - fpsync: mask broken pipe errors when dequeueing jobs (fixes GH issue #22)
    - fpsync: fix bug where a job could remain resumable forever under certain
      conditions.
    - fpsync: fix detection of fpart exiting with errors
    - fpsync: add prepare run option (-p)
    - fpsync: add list runs options (-l)
    - fpsync: add replay run option (-R)
    - fpsync: add archive run option (-a)
      Note for packagers: adds a new dependency to tar(1)
    - fpsync: add delete run option (-D)
    - fpsync: rework help and man page for clarification
    - fpsync: add check for incompatible rsync option (--delete)
    - fpsync: add send mail option (-M)
      Note for packagers: adds a new dependency to mail(1)
2020/01/10, 1.2.0 ('Joking') :
    - fpsync: ability to set an unlimited number of files or an unlimited size
      per sync job (fixes GH issue #9)
    - fpart: exclude files earlier (fixes GH issue #14)
      Before that change, excluded files would not make a directory empty and
      still account for its size; excluded files are now considered absent.
      That change does not apply when a directory size is computed after
      a certain depth (option -d) has been reached (in that case, once the
      maximum depth has been reached, every single file accounts for a directory
      size).
    - fpart: allow to specify paths with options '-y', '-Y', '-x' and '-X'
      (fixes GH issue #17)
    - fpart: raise limits for different internal types: the size of a file or
      partition, the number of files inside a partition and the number of
      partitions. That will make fpart(1) ready to handle bigger filesystems.
    - fpart: accept human-friendly size formats for options -s, -p, -q and -r
    - fpsync: accept human-friendly size formats for option -s
    - fpsync: use 'command' instead of 'which' to find binaries
      (fixes GH issue #20)
2018/11/15, 1.1.0 ('Reading') :
    - fpart: options -D and -E now pack files if they are explicitly passed as
      arguments
    - fpsync: simplify and fix detection of remote URLs (fixes GH issue #2)
    - fpsync: implement simple ETA when displaying status
    - fts.h: portability fix for Solaris 10 and earlier (GH PR #4)
    - fpart: fix build on Alpine Linux (GH PR #5)
    - fpsync: add timestamps and part numbers to logs
    - fpsync: add support for cpio copy tool (option -m)
    - fpart: option -Z becomes option -zz (incremental -z)
    - fpart: add option -zzz to add all directories to file lists (as empty)
    - other: lower required autoconf version to 2.63 (GH issue #7)
    - other: update RPM .spec file for CentOS and Feroda (GH issue #7)
2017/11/10, 1.0.0 ('Dancing'):
    - fpart: update embedded fts(3) using FreeBSD 12-CURRENT's version
    - fpart: add option -E to pack directories instead of files
    - fpsync: add option -E to work with directories (enables rsync(1)'s --delete)
      That option finally enables fpsync(1) to parallelize the so-called final
      rsync(1) pass, needed to remove extra files with the --delete option!
      (of course it can also be used on a regular basis instead of the usual
      'file' mode: you will get a perfect mirror without needing any extra
      --delete pass)
    - fpart: add option -0 to end filenames with a null (\0) character when
      using option -o
    - fpsync: use fpart's new option -0 in conjunction with rsync's --from0 option
      to avoid failing on filenames with special characters
    - fpsync: improve documentation
2017/04/27, 0.9.3 ('Funny words'):
    - fpsync: add support for rsync URL as target
      Suggested and tested by Harry Mangalam. Thanks, Harry!
2015/02/19, 0.9.2 ('What's that?'):
    - fpsync: add option '-S' to use sudo(1) for filesystem crawling and synchronizations
    - fpsync: add option '-O' to override default fpart(1) options
    - add RPM .spec file (see the contribs/package/rpm directory)
2015/02/06, 0.9.1 ('Let's play together'):
    - add tools/fpsync: a tool to sync directories in parallel using fpart and
      rsync. See fpsync(1) for more details.
    - fpart: print the number of files found in verbose mode only
2013/11/13, 0.9-1:
    (this minor change only impacts tarball users, fpart code itself has not
    changed)
    - Backport the following patch to config.guess:
      http://git.savannah.gnu.org/gitweb/?p=config.git;a=commitdiff;h=29900d3bff1ce445087ece5cb2cac425df1c2f74
      it adds support for ppcle and ppc64le architectures.
    Submitted by: Madhu Pavan <kmp@linux.vnet.ibm.com>
2013/09/10, 0.9 ('I bite!'):
    - Code cleanup
    - Fix Debian bug #719338 (fix build on 32bit architectures)
2013/06/25, 0.8 ('Moving around'):
    - Keep environ(7) when forking hooks
    - Add sections about live mode and data migration in README
    - Use autotools and get rid of manual Makefiles
2013/02/18, 0.7 ('Can I take it ?'):
    - Added option -D to group leaf directories as single file entries
    - Big options and doc update/cleanup
    - Sync'd fts(3) code with current FreeBSD version (svn rev 245505)
    - Embedded fts(3) can now be used on GNU/Linux
    - Renamed option '-x' to '-b'
    - Added options '-y', '-Y', '-x' and '-X' to include or exclude files
2013/01/21, 0.6 ('Very funny'):
    - Options handling cleanup and various bugfixes

examples/fpsync.sh
#!/bin/sh

# Copyright (c) 2014 Ganael LAPLANCHE <ganael.laplanche@martymac.org>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

# This script is a simple wrapper showing how fpart can be used to migrate data.
# It uses fpart and rsync to spawn multiple rsync instances to migrate data
# from src_dir/ to dst_dir/.
# This migration is incremental and will need a final rsync pass
# (rsync -av -delete src_dir/ dst_dir/) to remove extra files from dst_dir/.

########## Helper functions

usage () {
    echo "$0 <src_dir> <dst_dir>"
}

echo_log () {
    [ -n "$1" ] && echo "$1" | \
        tee -a "${FPART_LOGFILE}"
}

end_die () {
    [ -n "$1" ] && echo -e "$1" 1>&2
    exit 1
}

is_null () {
    echo "$1" | grep -qE "^[[:space:]]*$"
}

# Handle ^C : stop queue processing by setting the "stop" flag
trap_queue () {
    job_queue_stop
    end_die "\nKilled."
}
trap 'trap_queue' 2

########## Needed for config set-up

# Check for args presence
if is_null "$1" || is_null "$2"
then
    usage
    end_die
fi
SRC_PATH="$1"
DST_PATH="$2"

# Read the job name
while read -p "Enter a job name: " FPART_JOBNAME && \
    echo "${FPART_JOBNAME}" | grep -vqE '^[a-zA-Z0-9]+$'
do
    :
done

########## Config

# Queue manager configuration. This queue remains local, even when using SSH.
JOBS_MAX=4                                              # How many jobs to run
#JOBS_MAX=$(($(sysctl -n hw.ncpu) - 1))
JOBS_QUEUEDIR="/var/tmp/fpart/queue/${FPART_JOBNAME}"   # Queue dir.
JOBS_WORKDIR="/var/tmp/fpart/work/${FPART_JOBNAME}"     # Current jobs' dir.

# Paths for executables
FPART_BIN="/usr/local/bin/fpart"
RSYNC_BIN="/usr/local/bin/rsync"
SSH_BIN="/usr/bin/ssh"

# Fpart paths. Those ones must be shared amongst all nodes when using SSH.
FPART_LOGDIR="/mnt/nfsshared/fpart/log/${FPART_JOBNAME}"
FPART_LOGFILE="${FPART_LOGDIR}/fpart.log"
FPART_OUTPARTDIR="/mnt/nfsshared/fpart/partitions/${FPART_JOBNAME}"

# Fpart options
FPART_MAXPARTFILES="20000"
FPART_MAXPARTSIZE="$((10 * 1024 * 1024 * 1024))" # 10 GB
FPART_OPTIONS="-x '.zfs' -x '.snapshot*'"

# Fpart hooks
FPART_COMMAND="/bin/sh -c '${RSYNC_BIN} -av --numeric-ids \
        --files-from=\\\"\${FPART_PARTFILENAME}\\\" \
        \\\"${SRC_PATH}/\\\" \
        \\\"${DST_PATH}/\\\"' \
        1>\"${FPART_LOGDIR}/$$-\${FPART_PARTNUMBER}.stdout\" \
        2>\"${FPART_LOGDIR}/$$-\${FPART_PARTNUMBER}.stderr\""
FPART_POSTHOOK="echo \"${FPART_COMMAND}\" > \
        \"${JOBS_QUEUEDIR}/\${FPART_PARTNUMBER}\" && \
        echo \"=> [FPART] Partition \${FPART_PARTNUMBER} written\"" # [1]

# [1] Be careful to host the job queue on a filesystem that can handle
# fine-grained mtime timestamps (i.e. with a sub-second prescision) if you want
# the queue to be processed in order when fpart generates several job files per
# second.
# On FreeBSD, the precision can be tuned using the vfs.timestamp_precision
# sysctl. See vfs_timestamp(9).

# Mail
MAIL_ADDR="storage@mydomain.mytld"

# SSH hosts (comment out to execute jobs locally)
#SSH_HOSTS="login@host1 login@host2 login@host3"

########## Jobs-related functions

# Initialize job queue and work directories
init_job_queue () {
    mkdir -p "${JOBS_QUEUEDIR}" 2>/dev/null || \
        end_die "Cannot create job queue directory ${JOBS_QUEUEDIR}"
    rm -f "${JOBS_QUEUEDIR}"/*
    mkdir -p "${JOBS_WORKDIR}" 2>/dev/null || \
        end_die "Cannot create job work directory ${JOBS_WORKDIR}"
    rm -f "${JOBS_WORKDIR}"/*
}

# Set the "done" flag within job queue
job_queue_done () {
    sleep 1 # Ensure this very last file gets created within the next second of
            # last job file's mtime. Necessary for filesystems that don't get
            # below the second for mtime precision (msdosfs).
    touch "${JOBS_QUEUEDIR}/done"
}

# Set the "stop" flag within job queue
job_queue_stop () {
    touch "${JOBS_QUEUEDIR}/stop"
}

# Get next job name relative to ${JOBS_WORKDIR}/
# Returns empty string if no job is available
dequeue_job () {
    local _NEXT=""
    if [ -f "${JOBS_QUEUEDIR}/stop" ]
    then
        echo "stop"
    else
        _NEXT=$(cd "${JOBS_QUEUEDIR}" && ls -rt1 | head -n 1)
        if [ -n "${_NEXT}" ]
        then
            mv "${JOBS_QUEUEDIR}/${_NEXT}" "${JOBS_WORKDIR}" || \
                end_die "Cannot dequeue next job"
            echo "${_NEXT}"
        fi
    fi
}

# Push a PID to the list of currently-running-PIDs
push_pid_queue () {
    if [ -n "$1" ]
    then
        JOBS_QUEUE="${JOBS_QUEUE} $1"
        JOBS_NUM="$((${JOBS_NUM} + 1))"
    fi
}

# Rebuild the currently-running-PIDs' list by examining each process' state
refresh_pid_queue () {
    local _JOBS_QUEUE=""
    local _JOBS_NUM=0
    for _PID in ${JOBS_QUEUE}
    do
        if ps "${_PID}" 1>/dev/null 2>&1
        then
            _JOBS_QUEUE="${_JOBS_QUEUE} ${_PID}"
            _JOBS_NUM="$((${_JOBS_NUM} + 1))"
        fi
    done
    JOBS_QUEUE=${_JOBS_QUEUE}
    JOBS_NUM=${_JOBS_NUM}
}

# Simple round-robin for SSH_HOSTS
rotate_ssh_hosts () {
    if [ -n "${SSH_HOSTS}" ]
    then
        SSH_HOSTS="$(echo ${SSH_HOSTS} | cut -d ' ' -f 2-) $(echo ${SSH_HOSTS} | cut -d ' ' -f 1)"
    fi
}

# Pick-up next SSH host
next_ssh_host () {
    if [ -n "${SSH_HOSTS}" ]
    then
        echo "${SSH_HOSTS}" | cut -d ' ' -f 1
    fi
}

# Main jobs loop : pick up jobs within the queue directory and start them
jobs_loop () {
    local _NEXT=""
    while [ "${_NEXT}" != "done" ] && [ "${_NEXT}" != "stop" ]
    do
        local _PID=""
        if [ ${JOBS_NUM} -lt ${JOBS_MAX} ]
        then
            _NEXT="$(dequeue_job)"
            if [ -n "${_NEXT}" ] && [ "${_NEXT}" != "done" ] && [ "${_NEXT}" != "stop" ]
            then
                echo_log "=> [QMGR] Starting job ${JOBS_WORKDIR}/${_NEXT}"
                if [ -z "${SSH_HOSTS}" ]
                then
                    /bin/sh "${JOBS_WORKDIR}/${_NEXT}" &
                else
                    "${SSH_BIN}" "$(next_ssh_host)" '/bin/sh -s' \
                        < "${JOBS_WORKDIR}/${_NEXT}" &
                    rotate_ssh_hosts
                fi
                push_pid_queue $!
            fi
        fi
        refresh_pid_queue
        sleep 0.2
    done

    if [ "${_NEXT}" == "done" ]
    then
        echo_log "=> [QMGR] Done submitting jobs. Waiting for them to finish."
    else
        echo_log "=> [QMGR] Stopped. Waiting for jobs to finish."
    fi
    wait
    echo_log "=> [QMGR] No more job running."
}

########## Program start

FPART_OUTPARTTEMPL="${FPART_OUTPARTDIR}/part-$$"

JOBS_NUM=0      # Current num of concurrent processes
JOBS_QUEUE=""   # Jobs PID queue

# Checks for binaries
if [ ! -x "${FPART_BIN}" ] || [ ! -x "${RSYNC_BIN}" ] || [ ! -x "${SSH_BIN}" ]
then
    end_die "External tools are missing, check your configuration"
fi

# Create fpart directories and log file
mkdir -p "${FPART_OUTPARTDIR}" 2>/dev/null || \
    end_die "Cannot create output directory: ${FPART_OUTPARTDIR}"
mkdir -p "${FPART_LOGDIR}" 2>/dev/null || \
    end_die "Cannot create log directory: ${FPART_LOGDIR}"
touch "${FPART_LOGFILE}" 2>/dev/null || \
    end_die "Cannot create log file: ${FPART_LOGFILE}"

# Validate src and dst paths
if [ -n "${SSH_HOSTS}" ]
then
    # When using SSH, src/ path must exist locally (to allow fpart crawling it)
    # and src/ and dst/ paths must exist on each node.
    [ ! -d "${SRC_PATH}" ] && \
        end_die "Directory ${SRC_PATH} does not exist locally (or is not a directory)"
    # This also allows checking the SSH connectivity to each declared host.
    echo_log "======> Validating src/ and dst/ paths on SSH nodes..."
    for _host in ${SSH_HOSTS}
    do
        "${SSH_BIN}" "${_host}" "/bin/sh -c '[ -d "${SRC_PATH}" ]'" || \
            end_die "Directory ${SRC_PATH} does not exist for target ${_host} (or is not a directory)"
        "${SSH_BIN}" "${_host}" "/bin/sh -c '[ -d "${DST_PATH}" ]'" || \
            end_die "Directory ${DST_PATH} does not exist for target ${_host} (or is not a directory)"
    done
else
    [ ! -d "${SRC_PATH}" ] && \
        end_die "Directory ${SRC_PATH} does not exist (or is not a directory)"
    [ ! -d "${DST_PATH}" ] && \
        end_die "Directory ${DST_PATH} does not exist (or is not a directory)"
fi

# Initialize jobs queue and start jobs_loop
init_job_queue
jobs_loop&

# Let's rock !
echo_log "======> [$$] Syncing ${SRC_PATH} => ${DST_PATH}"
echo_log "===> Start time : $(date)"
echo_log "===> Starting fpart..."
echo_log "===> (parts dir : ${FPART_OUTPARTDIR})"
echo_log "===> (log dir : ${FPART_LOGDIR})"

# Start fpart from src_dir/ directory and produce jobs within ${JOBS_QUEUEDIR}/
cd "${SRC_PATH}" && \
    time "${FPART_BIN}" -f "${FPART_MAXPARTFILES}" -s "${FPART_MAXPARTSIZE}" \
        -o "${FPART_OUTPARTTEMPL}" ${FPART_OPTIONS} -Z -L \
        -W "${FPART_POSTHOOK}" . 2>&1 | tee -a "${FPART_LOGFILE}"

# Tell jobs_loop that crawling has finished
job_queue_done

# Wait for jobs_loop to terminate
echo_log "===> Jobs submitted, waiting..."
wait

# Examine results and send e-mail if requested
RET=$(find "${FPART_LOGDIR}/" -name "$$-*.stderr" ! -size 0)
MSG=$(echo -e "Return code: $( ([ -z "${RET}" ] && echo '0') || echo '1')" ; \
    [ -n "${RET}" ] && echo -e "Rsync errors detected, see logs:\n${RET}")
if [ -n "${MAIL_ADDR}" ]
then
    echo "${MSG}" | mail -s "Fpart job ${FPART_JOBNAME}" "${MAIL_ADDR}"
fi
echo_log "===> Jobs terminated."
echo_log "${MSG}"
echo_log "<=== End time : $(date)"

[ -n "${RET}" ] && exit 1
exit 0
